core/drand_daemon_control.go

```diff
@@ -33,7 +33,13 @@ func (dd *DrandDaemon) InitDKG(c context.Context, in *drand.InitDKGPacket) (*dra
 		}
 	}
 
-	return bp.InitDKG(c, in)
+	chainGroup, err := bp.InitDKG(c, in)
+	if err == nil {
+		// Add beacon handler from chain has for http server
+		dd.AddBeaconHandler(beaconID, bp)
```

Comment:
I think my answer would be useful for both points. We have to split the places because they are meant to do different things. When we first load the daemon, we read from disk all the beacon processes present. Some of them may have run before while others are waiting to run DKG. Both of them should be available to receive incoming messages from others nodes and possible clients coming from the daemon. When we really check that the beacon have run before and we have its chain hash, we add it to the http handler. I understand that DKG could not have ended successfully, but it is a lot harder to register the http handler on that callback than where we decided to do it.

Reply:
I dont think it's that hard. You need to change that [allback](https://github.com/drand/drand/blob/master/core/config.go#L184)  to take a share and the group, so you can derive the group info. This callback is now irrelevant in its current form because we don't know for which DKG it is being called so we need to fix it either way.
And like I wrote, currently this code is not functioning properly because client can kill the command but the DKG can still finish, so it's not gonna get registered on the http handler.

Reply:
> And like I wrote, currently this code is not functioning properly because client can kill the command but the DKG can still finish, so it's not gonna get registered on the http handler.

It doesn't matter if the user kills the command, as the logic happens inside drand daemon. This will be executed anyway, won't it?

```
chainGroup, err := bp.InitDKG(c, in)
if err == nil {
  // Add beacon handler from chain hash for http server
  dd.AddBeaconHandler(beaconID, bp)
}

return chainGroup, err
```

I need the entire reference to the beacon process, not only the group, as you can see [here](https://github.com/drand/drand/blob/4137837c458b80df05555f5197dc4666bdb8ddc4/core/drand_daemon.go#L148). Do you think it is correct to pass it on the callback? I really don't like how that looks like.

Reply:
It is a problem because this logic happens in the gRPC call, and it will return early if the client kills the connection/CTRL-C, because you pass the context in the `bp.InitDKG` call, so it will return early _even if the DKG is still continuing_ and you'll not get the chance to register the handler. Does that make sense ?
I dont think you need the reference for the beacon process as you can get it from the chain hash at the drand daemon level, and if the callback gives you the group you can get the chain info from which you can get the chain hash.

Reply:
> It is a problem because this logic happens in the gRPC call, and it will return early if the client kills the connection/CTRL-C, because you pass the context in the bp.InitDKG call, so it will return early even if the DKG is still continuing and you'll not get the chance to register the handler. Does that make sense ?
I think that makes sense. I have changed the way we register new beacon handlers on http server. Please check it now!